lib/active_merchant/billing/gateways/shift4.rb

```diff
@@ -0,0 +1,273 @@
+module ActiveMerchant #:nodoc:
+  module Billing #:nodoc:
+    class Shift4Gateway < Gateway
+      self.test_url = 'https://utgapi.shift4test.com/api/rest/v1/'
+      self.live_url = 'https://utg.shift4api.net/api/rest/v1/'
+
+      self.supported_countries = ['US']
+      self.default_currency = 'USD'
+      self.supported_cardtypes = %i[visa master american_express discover]
+
+      self.homepage_url = 'https://shift4.com'
+      self.display_name = 'Shift4'
+
+      STANDARD_ERROR_CODE_MAPPING = {
+        'incorrect_number' => STANDARD_ERROR_CODE[:incorrect_number],
+        'invalid_number' => STANDARD_ERROR_CODE[:invalid_number],
+        'invalid_expiry_month' => STANDARD_ERROR_CODE[:invalid_expiry_date],
+        'invalid_expiry_year' => STANDARD_ERROR_CODE[:invalid_expiry_date],
+        'invalid_cvc' => STANDARD_ERROR_CODE[:invalid_cvc],
+        'expired_card' => STANDARD_ERROR_CODE[:expired_card],
+        'insufficient_funds' => STANDARD_ERROR_CODE[:card_declined],
+        'incorrect_cvc' => STANDARD_ERROR_CODE[:incorrect_cvc],
+        'incorrect_zip' => STANDARD_ERROR_CODE[:incorrect_zip],
+        'card_declined' => STANDARD_ERROR_CODE[:card_declined],
+        'processing_error' => STANDARD_ERROR_CODE[:processing_error],
+        'lost_or_stolen' => STANDARD_ERROR_CODE[:card_declined],
+        'suspected_fraud' => STANDARD_ERROR_CODE[:card_declined],
+        'expired_token' => STANDARD_ERROR_CODE[:card_declined]
+      }
+
+      def initialize(options = {})
+        requires!(options, :access_token)
```

Comment:
It looks like this has changed, but it looks like we need `client_guid` and `auth_token` in order to retrieve this `access_token`?

In the docs for [Access Token Exchange](https://myportal.shift4.com/index.cfm?action=development.shift4api#section/Development-Quick-Start/Access-Token-Exchange), we should require those 2 values here and execute the request to exchange those for an access token that we then store with the gateway in Core.

Take a look at the `initialize` method in [Airwallex](https://github.com/activemerchant/active_merchant/blob/023f3e8685e763c8758292790e2962dbaffc0c1b/lib/active_merchant/billing/gateways/airwallex.rb#L35). In the constructor, we make a call to the required endpoint to generate this value, that we in turn store on our end whenever a customer creates a gateway.

The request is defined in the [`setup_access_token` method](https://github.com/activemerchant/active_merchant/blob/023f3e8685e763c8758292790e2962dbaffc0c1b/lib/active_merchant/billing/gateways/airwallex.rb#L128-L141) which I think you can probably copy over and be able to get it to work here.

It also notes in the documentation that the `auth_token` is discarded after the call to retrieve the `access_token`, so we may need to consider how to handle that scenario so that users are re-creating Shift4 gateways in our system with a bunch of `auth_token`s.